internal/executor/job/processors/remove_runs.go

```diff
@@ -0,0 +1,69 @@
+package processors
+
+import (
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+
+	executorContext "github.com/armadaproject/armada/internal/executor/context"
+	"github.com/armadaproject/armada/internal/executor/domain"
+	"github.com/armadaproject/armada/internal/executor/job"
+	"github.com/armadaproject/armada/internal/executor/util"
+)
+
+type RemoveRunProcessor struct {
+	clusterContext   executorContext.ClusterContext
+	jobRunStateStore job.RunStateStore
+}
+
+func NewRemoveRunProcessor(clusterContext executorContext.ClusterContext, jobRunStateStore job.RunStateStore) *RemoveRunProcessor {
+	return &RemoveRunProcessor{
+		clusterContext:   clusterContext,
+		jobRunStateStore: jobRunStateStore,
+	}
+}
+
+func (j *RemoveRunProcessor) Run() {
+	managedPods, err := j.clusterContext.GetBatchPods()
+	if err != nil {
+		log.Errorf("Failed to cancel runs because unable to get a current managed pods due to %s", err)
+		return
+	}
+
+	runsToCancel := j.jobRunStateStore.GetAllWithFilter(func(state *job.RunState) bool {
+		return state.CancelRequested
+	})
+	runPodInfos := createRunPodInfos(runsToCancel, managedPods)
+
+	util.ProcessItemsWithThreadPool(context.Background(), 20, runPodInfos,
+		func(runInfo *runPodInfo) {
+			pod := runInfo.Pod
+			if pod == nil {
+				j.jobRunStateStore.Delete(runInfo.Run.Meta.RunId)
+				return
+			}
+
+			if util.IsInTerminalState(pod) && util.HasCurrentStateBeenReported(pod) {
```

Comment:
Ho good is the state model here?  e.g  if a pod is simultaneously CancelRequested and PreemptionRequested, what sort of guarantees do we get about what will happen?

Reply:
Still relatively weak - it isn't a unified state yet

So no guarantee - beyond the pod will get deleted 

It is a fair point - we could run these sequentially rather in parallel (or combine them)